### PR TITLE
Docs: Add code snippet to addons panel

### PR DESCRIPTION
**Pull request by larsrickert**
Closes #26691

## What I did

Add a new "Code" panel for the docs addon so the generated code snippet can also be viewed/copied in single story mode and is no longer limited to the docs page / view mode.

image

### Open questions:

- Since the code snippet generation is way more advanced for Vue than for other frameworks (when using vue-component-meta docgen), should be provide a preview option to hide/disable the "Code" panel?
- Should we remove the top/bottom margin of the source code?

#### The changes in this PR are covered in the following automated tests:
- [x] stories
- [ ] unit tests
- [ ] integration tests
- [ ] end-to-end tests

#### Manual testing

1. Run a sandbox for template, e.g. `yarn task --task sandbox --start-from auto --template vue3-vite/default-ts`
2. Open Storybook in your browser
3. Access any single story, e.g. http://localhost:6006/?path=/story/stories-renderers-vue3-vue3-vite-default-ts-sourcecode--default
4. Click on "Code" panel at the bottom of the screen

### Documentation

- [ ] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [x] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli/src/sandbox-templates.ts`
- [x] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

     - `bug`: Internal changes that fixes incorrect behavior.
     - `maintenance`: User-facing maintenance tasks.
     - `dependencies`: Upgrading (sometimes downgrading) dependencies.
     - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
     - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
     - `documentation`: Documentation **only** changes. Will not show up in release changelog.
     - `feature request`: Introducing a new feature.
     - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
     - `other`: Changes that don't fit in the above categories.
   
   </details>

### 🦋 Canary release

This PR does not have a canary release associated. You can request a canary release of this pull request by mentioning the `@storybookjs/core` team here.

_core team members can create a canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=<PR_NUMBER>`_

| name | before | after | diff | z | % |
| ---- | ------ | ----- | ---- | - | - |
| createSize |  0 B | 0 B | 0 B | - | - |
| generateSize |  77.7 MB | 77.7 MB | 0 B | 0.58 | 0% |
| initSize |  133 MB | 133 MB | 3.35 kB | 0.68 | 0% |
| diffSize |  55.3 MB | 55.3 MB | 3.35 kB | 0.68 | 0% |
| buildSize |  6.87 MB | 7.24 MB | 363 kB | **1448.14** | 🔺5% |
| buildSbAddonsSize |  1.51 MB | 1.88 MB | 363 kB | **1435.12** | 🔺19.3% |
| buildSbCommonSize |  195 kB | 195 kB | 0 B | - | 0% |
| buildSbManagerSize |  1.86 MB | 1.86 MB | 75 B | **2.34** | 0% |
| buildSbPreviewSize |  0 B | 0 B | 0 B | - | - |
| buildStaticSize |  0 B | 0 B | 0 B | - | - |
| buildPrebuildSize |  3.57 MB | 3.93 MB | 363 kB | **1459.31** | 🔺9.2% |
| buildPreviewSize |  3.3 MB | 3.3 MB | 168 B | **7.78** | 0% |
| testBuildSize |  0 B | 0 B | 0 B | - | - |
| testBuildSbAddonsSize |  0 B | 0 B | 0 B | - | - |
| testBuildSbCommonSize |  0 B | 0 B | 0 B | - | - |
| testBuildSbManagerSize |  0 B | 0 B | 0 B | - | - |
| testBuildSbPreviewSize |  0 B | 0 B | 0 B | - | - |
| testBuildStaticSize |  0 B | 0 B | 0 B | - | - |
| testBuildPrebuildSize |  0 B | 0 B | 0 B | - | - |
| testBuildPreviewSize |  0 B | 0 B | 0 B | - | - |

| name | before | after | diff | z | % |
| ---- | ------ | ----- | ---- | - | - |
| createTime |  21s | 6.6s | -14s -439ms | -0.97 | -218.1% |
| generateTime |  20.1s | 21.4s | 1.2s | -0.26 | 6% |
| initTime |  14.3s | 12.6s | -1s -681ms | -0.84 | -13.3% |
| buildTime |  9.9s | 9.7s | -131ms | 1.16 | -1.3% |
| testBuildTime |  0ms | 0ms | 0ms | - | - |
| devPreviewResponsive |  4.4s | 4.3s | -72ms | -1.13 | -1.6% |
| devManagerResponsive |  3.2s | 3.3s | 109ms | -1.03 | 3.3% |
| devManagerHeaderVisible |  488ms | 514ms | 26ms | -0.69 | 5.1% |
| devManagerIndexVisible |  571ms | 590ms | 19ms | -0.61 | 3.2% |
| devStoryVisibleUncached |  1.1s | 1.1s | 64ms | **-1.99** | 🔺5.3% |
| devStoryVisible |  571ms | 543ms | -28ms | -1.02 | -5.2% |
| devAutodocsVisible |  473ms | 445ms | -28ms | -0.79 | -6.3% |
| devMDXVisible |  436ms | 434ms | -2ms | -0.82 | -0.5% |
| buildManagerHeaderVisible |  483ms | 493ms | 10ms | -0.9 | 2% |
| buildManagerIndexVisible |  570ms | 582ms | 12ms | -0.91 | 2.1% |
| buildStoryVisible |  448ms | 459ms | 11ms | -0.8 | 2.4% |
| buildAutodocsVisible |  368ms | 418ms | 50ms | -0.43 | 12% |
| buildMDXVisible |  378ms | 385ms | 7ms | -0.72 | 1.8% |

## Greptile Summary

This PR adds a new 'Code' panel to the Storybook UI for displaying and copying generated code snippets in single story view mode, focusing on improvements for the docs addon and Vue3 renderer.

- Implemented new 'Code' panel in `code/addons/docs/src/manager.tsx` for viewing generated code snippets
- Enhanced Vue3 source code generation in `code/renderers/vue3/src/docs/sourceDecorator.ts`, improving handling of complex props and slots
- Updated `code/addons/essentials/package.json` and related files to integrate the new 'Code' panel functionality
- Reordered addons in `code/addons/essentials/src/index.ts`, potentially affecting addon initialization order
- Removed view mode check in Vue3 sourceDecorator, allowing source code generation for non-docs view modes

### CODE REVIEW

File: code/addons/essentials/src/index.ts
Comment (chakAs3) [inline code/addons/essentials/src/index.ts:93]: This will be applied to all the renderers, is there any reason why you changed the order? it is important

File: code/renderers/vue3/src/docs/sourceDecorator.ts
Comment (chakAs3) [inline code/renderers/vue3/src/docs/sourceDecorator.ts:128]: I think showing this on story view. mode should be opt-in, so you add new field the docs parameter

**Review by chakAs3 — CHANGES_REQUESTED** [review]

**Comment by chakAs3** [comment]
Hey @larsrickert, I like how you used the same code snippet on the SourceDecorator. However, this could ideally be an addon, similar to the existing StorySource addon. Though StorySource serves a different purpose—it only displays the actual story code from the file.

File: code/addons/essentials/src/index.ts
Comment (larsrickert) [reply code/addons/essentials/src/index.ts:93]: I just wanted to show the Code panel after controls and actions so it is not the first panel.
Is there any other way to achieve this?

File: code/renderers/vue3/src/docs/sourceDecorator.ts
Comment (larsrickert) [reply code/renderers/vue3/src/docs/sourceDecorator.ts:128]: For all other renderers it is always generated, we currently only check for the view mode for Vue.
Also in my opinion, the Code panel is a big benefit for users. What do you think about adding an opt-out preview parameter so the tab can be disabled?

**Comment by larsrickert** [comment]
> Hey @larsrickert, I like how you used the same code snippet on the SourceDecorator. However, this could ideally be an addon, similar to the existing StorySource addon. Though StorySource serves a different purpose—it only displays the actual story code from the file.

@chakAs3 Thanks for the feedback! Imo the code generation / snippets are responsibility of the docs addon which also already shows the snippets on the "Docs" page. So this PR/feature "just" also shows the snippets on the single story pages.
So its more like adding a missing feature / extending an existing feature of the docs addon instead of creating a new addon/feature.

File: code/renderers/vue3/src/docs/sourceDecorator.ts
Comment (chakAs3) [reply code/renderers/vue3/src/docs/sourceDecorator.ts:128]: I agree that the Code panel is a good addition and will be beneficial for users. Since Docs is an opt-in feature, I'll leave it to @shilman to decide on that part. I'm okay with everything else in the PR.

**Review by chakAs3 — APPROVED** [review]

**Comment by larsrickert** [comment]
@shilman Hey :) Would it be possible for you to take a look at this PR?

**Comment by shilman** [comment]
@larsrickert Sorry I missed this one. Will look later today!

**Review by shilman — COMMENTED** [review]
Hi @larsrickert ! Thanks for your patience on the review.

This looks great: a low-cost improvement to Storybook, eliminating the need for the Storysource addon, which is broken/outdated.

My original thought was to create a standalone addon for this, `@storybook/addon-source`, which would co-exist with `addon-docs` the same way that `@storybook/addon-controls` current does. What do you think about that direction?

I realize this is a much simpler change, and I appreciate that.

I want to discuss this with the core team on Monday and get back to you. But I like this direction and appreciate your PR!!!

**Comment by shilman** [comment]
@larsrickert I haven't had a formal discussion with the team but initial sentiment from talking to a few people is that we should move controls & source handling into the core, and keep docs as its own addon. I will keep you posted on whether we're serious about that change & how it might get done.

In the meantime, I'm inclined to merge this PR since it's useful as is. I will try to have some resolution this week.

**Comment by larsrickert** [comment]
@shilman Thanks for your insights and thoughts! In general, I appreciate the separation of concerns by moving features to dedicated addons. In the case of the docs addon / source generation I am open to both options. Since the source code generation is already part of the docs addon (and its already displayed on the "Docs" page for each story), the new "Code" panel would also fit perfectly in the current docs addon in my opinion.

I like your idea about merging this feature already in the docs addon and then discuss if it should be split up into a source addon in the future. But since this would be a bigger refactoring and probably also other parts of the docs addon would be moved, I'd still merge this PR already as feature for Storybook 8.x. 👍

**Comment by shilman** [comment]
Hi @larsrickert, can you please add the following flag:

```js
parameters: {
  docs: {
    source: {
       addonPanel: false
    },
  },
}
```

This way users can opt out of the new behavior if they want. Once that's ready and CI is green, I'll update `MIGRATION.md` to describe the new behavior and will merge the PR.

**Comment by larsrickert** [comment]
> Hi @larsrickert, can you please add the following flag:
> 
> ```js
> parameters: {
>   docs: {
>     source: {
>       addonPanel: false,
>     },
>   },
> }
> ```
> 
> This way users can opt out of the new behavior if they want. Once that's ready and CI is green, I'll update `MIGRATION.md` to describe the new behavior and will merge the PR.

Hey @shilman Sure, no problem :) Is there some example/docs how to implement this or someone that could guide me here? I already tried to achieve it but in `addons.add()` I can only hide the panel content by returning null, but the "Code" panel itself is still displayed. I guess we need to check the parameter before calling `addons.add`?

**Comment by shilman** [comment]
Oof. I'm actually not too familiar with that aspect of things. @ndelangen can you please help?

**Review by kasperpeulen — APPROVED** [review]
LGTM, and I agree with @shilman request to be able to turn it off.

**Comment by larsrickert** [comment]
@shilman @kasperpeulen Thanks for the feedback! I added the parameter as requested :)
Does someone of you have time to review if this is the correct way to do this?

**Comment by shilman** [comment]
@larsrickert Tested and it's working well. I'll take another pass at things tomorrow and try to get it merged. Great job!!

**Review by chakAs3 — APPROVED** [review]

**Comment by larsrickert** [comment]
> @larsrickert Tested and it's working well. I'll take another pass at things tomorrow and try to get it merged. Great job!!

@shilman Great, thanks for testing! Do you already have a rough release plan for this?

**Comment by shilman** [comment]
@larsrickert not so comfortable with the current parameter implementation. I have a cleaner solution with a worse user API. Need to figure that out, sorry for the delays. I'll update again later today

**Comment by shilman** [comment]
@larsrickert i discussed with @yannbf and @ndelangen and the current parameter approach won't cut it. i'm going to reimplement it now with a worse API and then get it merged. thanks again for the contribution and for your patience on this.

**Comment by shilman** [comment]
@larsrickert I've opened a PR on top of this one for your review. Please merge if you're OK with it, or we can discuss the changes there.

https://github.com/storybookjs/storybook/pull/29791

I've also found some issues with the PR in its current form.
- [ ] The default snippet is messed up for React out of the box (see sandbox)
- [ ] The panel should show up at the end so as not to disrupt existing setups

File: code/addons/docs/src/manager.tsx
Comment (chakAs3) [inline code/addons/docs/src/manager.tsx:14]: This is fabulous API 😆 . but still cleaner 👍

**Review by chakAs3 — APPROVED** [review]

File: MIGRATION.md
Comment (JReinhold) [inline MIGRATION.md:436]: Not sure I think this is the right API.

We already have `parameters.docs.source.X`: https://storybook.js.org/docs/api/doc-blocks/doc-block-source

Wouldn't it make sense to keep in line with that, and have it as `parameters.docs.source.panel = bool` instead?

**Review by JReinhold — CHANGES_REQUESTED** [review]
Great addition!

Don't we want to document this too?

File: MIGRATION.md
Comment (JReinhold) [reply MIGRATION.md:438]: I see that this has already been discussed, before.

Personally, I highly disagree with the direction.

**Review by shilman — APPROVED** [review]
LGTM. Will add some documentation in a followup PR 😬

**Comment by tomillie** [comment]
Any reason the Code panel doesn't take the `docs.source.format` and `docs.source.transform` configs into account?

**Comment by larsrickert** [comment]
> Any reason the Code panel doesn't take the `docs.source.format` and `docs.source.transform` configs into account?

No, I tried to consider the transformer but it seems like function parameters are not a available inside the addon because they are undefined when trying to use them. I'd need more time to investigate this

**Pull request merged**
